client/service/src/lib.rs

```diff
@@ -342,6 +364,29 @@ async fn build_network_future<
 			state_sink = status_sinks.state.next().fuse() => {
 				state_sink.send(network.network_state());
 			}
+
+			_ = (&mut handle_bitswap_events).fuse() => {}
+		}
+	}
+}
+
+/// Handle a `BitswapEvent`, reading and writing from the `OffchainStorage`.
+pub async fn handle_bitswap_event<B: BlockT, H: sc_network::ExHashT, S: IpldStore<Params=IpldStoreParams>>(
+	service: &Arc<sc_network::NetworkService<B, H>>,
+	event: sc_network::BitswapEvent,
+	storage: S,
+) -> Result<(), libipld::error::Error> {
+	match event {
+		sc_network::BitswapEvent::ReceivedBlock(_, cid, data) => {
+			storage.insert(libipld::Block::new(cid.clone(), data.to_vec())?).await?;
+			service.put_value(cid.to_bytes().into(), data.to_vec());
+			Ok(())
+		},
+		sc_network::BitswapEvent::ReceivedWant(peer_id, cid, _) => {
+			let block = storage.get(&cid).await?;
```

Comment:
get will potentially try to fetch the block from the network. I think you only want to respond if you have the block locally. If the block isn't available locally or found on the network it will return a `BlockNotFound`. In general there is a problem that the semantics of go-ipfs, js-ipfs are not clear (or outright insufficient), and while formulating and implementing precise semantics for ipfs-embed in terms of atomicity/durability/garbage collection, they are still in flux as more experience is gained building applications on top. That's why I'd recommend completely removing any storage logic for now.

Reply:
So should I just remove `handle_bitswap_event` and `IpldStoreBuilder` entirely?

Reply:
Not sure how you intend to use it. While exchanging blocks with full nodes and validators could be interesting if light clients can pay them to store blocks for them, for now I'm mainly interested in using it for light client to light client data exchange. So I think removing those entirely would be fine for now.